### ISSUE TITLE
A slice annotation to test Cassandra components

### ISSUE THREAD

**Issue by nosan** [spring-projects/spring-boot#16975]
`Spring Boot` does not provide a slice `@DataCassandraTest` annotation for test `Cassandra` **reactive** or **regular** components, as it does for `Redis`,`Mongo`, ...

The following snippet shows how it can be done:

```java

  import java.util.List;

  import com.github.nosan.embedded.cassandra.test.spring.Cql;
  import com.github.nosan.embedded.cassandra.test.spring.EmbeddedCassandra;
  import org.junit.Test;
  import org.junit.runner.RunWith;

  import org.springframework.beans.factory.annotation.Autowired;
  import org.springframework.boot.test.autoconfigure.data.cassandra.DataCassandraTest;
  import org.springframework.test.context.junit4.SpringRunner;

  import static org.assertj.core.api.Assertions.assertThat;

  @RunWith(SpringRunner.class)
  @DataCassandraTest
  @EmbeddedCassandra(scripts = "schema.cql")
  public class ExampleDataCassandraTests {

  	@Autowired
  	private ExampleRepository repository;

  	@Test
  	@Cql(scripts = "data.cql")
  	public void findAll() {
  		//
  	}

  }

```
By default, tests annotated with `@DataCassandraTest` will use an embedded Cassandra if either `@EmbeddedCassandra` annotation or
`Cassandra` bean is present in the `ApplicationContext`. If you don't want to use an embedded Cassandra, it will be a way to
exclude `TestCassandraAutoConfiguration` via `@DataCassandraTest.exclude()` attribute.

From other side it also will be able to use `@DataCassandra` annotation in conjunction with `test-container`, e.g.:

```java

import org.junit.jupiter.api.Test;
import org.testcontainers.containers.CassandraContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.docs.test.data.cassandra.ExampleRepository;
import org.springframework.boot.test.autoconfigure.data.cassandra.DataCassandraTest;
import org.springframework.boot.test.util.TestPropertyValues;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.test.context.ContextConfiguration;

@Testcontainers
@ContextConfiguration(initializers = DataCassandraTestContainerTests.Initializer.class)
@DataCassandraTest
class DataCassandraTestContainerTests {

	@Container
	public static final CassandraContainer<?> cassandra = new CassandraContainer<>("cassandra:3.11.4")
			.withInitScript("schema.cql");


	@Autowired
	private ExampleRepository repository;

	@Test
	void findAll() {
		//
	}

	static class Initializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {

		@Override
		public void initialize(ConfigurableApplicationContext context) {
			TestPropertyValues.of("spring.data.cassandra.port=" + cassandra.getFirstMappedPort())
					.applyTo(context.getEnvironment());
		}

	}

}

```

Currently, it looks a bit clumsy, but this [enhancement](https://github.com/spring-projects/spring-boot/issues/16886) can fix this in the future.

And finally `@DataCassandra` can be used with the `real` Cassandra which is running somewhere. Something like this:
```java

import org.junit.jupiter.api.Test;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.docs.test.data.cassandra.ExampleRepository;
import org.springframework.boot.test.autoconfigure.data.cassandra.DataCassandraTest;

@DataCassandraTest
class DataCassandraTestContainerTests {

	@Autowired
	private ExampleRepository repository;

	@Test
	void findAll() {
		//
	}

}


```

Summing up the above. It would be great to have such functionality,
as so far folks have to configure everything by themselves.

Of course if `Spring Boot Team` and `Community` are interested in it.

If yes, I'd be happy to implement this **feature**, otherwise feel free to close this issue.

Thanks in advance.

**Comment by snicoll** [comment]
@nosan thanks for the suggestion but I am a bit confused considering that you made a similar request that [was declined already](https://github.com/spring-projects/spring-boot/pull/13033#issuecomment-400430466). What am I missing?

**Comment by nosan** [comment]
Thank you for your feedback @snicoll.

You are right, it was declined last year as it was not ready at all. 
As I mentioned before, `@DataCassandraTest` can be used not only with an embedded Cassandra but with TestContainers or Сassandra itself. Still, from my perspective, it would be great if `Spring Boot` provides such functionality. 

In any case, if you still feel it is not necessary to have such a feature or it would be difficult to support it, feel free to close an issue. 
Thanks

**Comment by philwebb** [comment]
We discussed this today on our weekly call and we think it's a good idea to offer a `@DataCassandraTest` annotation. Our current feeling is that we should now be trying to encourage people to use TestContainers rather than embedded servers if possible. Given our time again, I'm not even sure we'd add embedded Mongo support.

We'll target this for 2.x because I think we should tackle it after we look at #16886

Thanks for suggesting it again @nosan!

**Comment by nosan** [comment]
Thank you very much.
If you need help let me know.

**Comment by wilkinsona** [comment]
Closing, somewhat tentatively, in favour of #17490.
